src/array_slice.cpp

```diff
@@ -0,0 +1,452 @@
+                                 field_function fun, void *fun_data,
+                                 bool has_imag,
+                                 void *vslice) {
+
+  am_now_working_on(ArraySlice);
+
+  /***************************************************************/
+  /* call get_array_slice_dimensions to get slice dimensions and */
+  /* partially initialze an array_slice_data struct              */
+  /***************************************************************/
+  int dims[3];
+  array_slice_data data;
+  int rank=get_array_slice_dimensions(where, dims, &data);
+  int slice_size=data.slice_size;
+  if (rank==0 || slice_size==0) return 0; // no data to write
+
+  cdouble *zslice;
+  double *slice;
+  if (vslice==0)
+   { if (has_imag)
+      { zslice = new cdouble[slice_size];
+        vslice = (void *)zslice;
+      } 
+     else
+      { slice  = new double[slice_size];
+        vslice = (void *)slice;
+      };
+   };
+   
+  data.vslice     = vslice;
+  data.has_imag   = has_imag;
+  data.fun        = fun ? fun : default_field_function;
+  data.fun_data   = fun_data;
+  data.components = components;
+
+  int num_components = components.size();
+
+  data.cS      = new component[num_components];
+  data.ph      = new cdouble[num_components];
+  data.fields  = new cdouble[num_components];
+  
+  data.offsets = new int[2 * num_components];
+  for (int i = 0; i < 2 * num_components; ++i)
+    data.offsets[i] = 0;
+
+  /* compute inverse-epsilon directions for computing Dielectric fields */
+  data.ninveps = 0;
+  bool needs_dielectric = false;
+  for (int i = 0; i < num_components; ++i)
+    if (components[i] == Dielectric) { needs_dielectric = true; break; }
+  if (needs_dielectric) 
+    FOR_ELECTRIC_COMPONENTS(c) if (gv.has_field(c)) {
+      if (data.ninveps == 3) abort("more than 3 field components??");
+      data.inveps_cs[data.ninveps] = c;
+      data.inveps_ds[data.ninveps] = component_direction(c);
+      ++data.ninveps;
+    }
+  
+  /* compute inverse-mu directions for computing Permeability fields */
+  data.ninvmu = 0;
+  bool needs_permeability = false;
+  for (int i = 0; i < num_components; ++i)
+    if (components[i] == Permeability) { needs_permeability = true; break; }
+  if (needs_permeability) 
+    FOR_MAGNETIC_COMPONENTS(c) if (gv.has_field(c)) {
+      if (data.ninvmu == 3) abort("more than 3 field components??");
+      data.invmu_cs[data.ninvmu] = c;
+      data.invmu_ds[data.ninvmu] = component_direction(c);
+      ++data.ninvmu;
+    }
+ 
+  loop_in_chunks(get_array_slice_chunkloop, (void *) &data,
+		 where, Centered, true, true);
+
+  /***************************************************************/
+  /* repeatedly call sum_to_all to consolidate full array slice  */
+  /* on all cores                                                */
+  /***************************************************************/
+#define BUFSIZE 1<<16 // use 64k buffer
+  if (has_imag)
+   { cdouble buffer[BUFSIZE];
```

Comment:
In general I would prefer not to stack-allocate here.  This is allocating 2¹⁶ * (16 bytes per `cdouble`) = 1MB for the buffer, which could easily overflow the stack size limit on some systems.  (e.g. the [default stack limit on Windows](https://software.intel.com/en-us/articles/memory-limits-applications-windows) is 1MB)

Use `new`/`delete` instead.